zerver/tests/test_message_move_topic.py

```diff
@@ -1784,3 +1785,119 @@ def test_resolve_empty_string_topic(self) -> None:
+        admins_group = NamedUserGroup.objects.get(
+            name=SystemGroups.ADMINISTRATORS, realm=admin_user.realm, is_system_group=True
+        )
+        do_change_realm_permission_group_setting(
+            admin_user.realm,
+            "can_resolve_topics_group",
+            admins_group.usergroup_ptr,
+            acting_user=None,
+        )
+
+        result = self.resolve_topic_containing_message(
+            hamlet,
+            id1,
+        )
+        self.assert_json_error(result, "You don't have permission to resolve topics.")
+
+        result = self.resolve_topic_containing_message(
+            admin_user,
+            id1,
+        )
+        self.assert_json_success(result)
+
+        # Test restrict resolving topics to a user defined group.
+        original_topic_name = "topic 2"
+        id2 = self.send_stream_message(hamlet, stream.name, topic_name=original_topic_name)
+        leadership_group = check_add_user_group(
+            admin_user.realm, "leadership", [hamlet], acting_user=hamlet
+        )
+        do_change_realm_permission_group_setting(
+            admin_user.realm, "can_resolve_topics_group", leadership_group, acting_user=None
+        )
+
+        result = self.resolve_topic_containing_message(
+            othello,
+            id2,
+        )
+        self.assert_json_error(result, "You don't have permission to resolve topics.")
+
+        result = self.resolve_topic_containing_message(
+            admin_user,
+            id2,
+        )
+        self.assert_json_error(result, "You don't have permission to resolve topics.")
+
+        result = self.resolve_topic_containing_message(
+            hamlet,
+            id2,
+        )
+        self.assert_json_success(result)
+
+        # Test restrict topics to an anonymous group.
+        original_topic_name = "topic 3"
+        id3 = self.send_stream_message(hamlet, stream.name, topic_name=original_topic_name)
+        staff_group = check_add_user_group(
+            admin_user.realm, "Staff", [admin_user], acting_user=admin_user
+        )
+        anonymous_setting_group = self.create_or_update_anonymous_group_for_setting(
+            [cordelia], [staff_group]
+        )
+        do_change_realm_permission_group_setting(
+            admin_user.realm, "can_resolve_topics_group", anonymous_setting_group, acting_user=None
+        )
+
+        result = self.resolve_topic_containing_message(
+            othello,
+            id3,
+        )
+        self.assert_json_error(result, "You don't have permission to resolve topics.")
+
+        result = self.resolve_topic_containing_message(
+            cordelia,
+            id3,
+        )
+        self.assert_json_success(result)
+
+        id3 = self.send_stream_message(hamlet, stream.name, topic_name=original_topic_name)
+        result = self.resolve_topic_containing_message(
+            admin_user,
+            id3,
+        )
+        self.assert_json_success(result)
```

Comment:
I feel like we may want to move away from having every permissions test check all the cases of how a group-setting value can be constructed, given how much logic they share. But that is mostly a thought for later refactoring.